services/horizon/internal/ingest/processors/asset_stats_set.go

```diff
@@ -379,9 +402,93 @@ func (s AssetStatSet) AddContractData(change ingest.Change) error {
 	return nil
 }
 
+func (s AssetStatSet) ingestAssetContractBalance(change ingest.Change) {
+	if change.Pre != nil {
+		contractID := change.Pre.Data.MustContractData().ContractId
+		holder, amt, ok := ContractBalanceFromContractData(*change.Pre, s.networkPassphrase)
+		if !ok {
+			return
+		}
+		stats, ok := s.contractAssetStats[contractID]
+		if !ok {
+			stats = contractAssetStatValue{
+				balance:    big.NewInt(0),
+				numHolders: 0,
+			}
+		}
+
+		if change.Post == nil {
+			// the balance was removed so we need to deduct from
+			// contract holders and contract balance amount
+			stats.balance = new(big.Int).Sub(stats.balance, amt)
+			// only decrement holders if the removed balance
+			// contained a positive amount of the asset.
+			if amt.Cmp(big.NewInt(0)) > 0 {
+				stats.numHolders--
+			}
+			s.maybeAddContractAssetStat(contractID, stats)
+			return
+		}
+		// if the updated ledger entry is not in the expected format then this
+		// cannot be emitted by the stellar asset contract, so ignore it
+		postHolder, postAmt, postOk := ContractBalanceFromContractData(*change.Post, s.networkPassphrase)
+		if !postOk || postHolder != holder {
+			return
+		}
+
+		delta := new(big.Int).Sub(postAmt, amt)
+		stats.balance = new(big.Int).Add(stats.balance, delta)
+		if postAmt.Cmp(big.NewInt(0)) == 0 && amt.Cmp(big.NewInt(0)) > 0 {
```

Comment:
I think you can use `delta` in place of `new(big.Int)` here to save an allocation? e.g.

```suggestion
		delta := new(big.Int).Sub(postAmt, amt)
		stats.balance = delta.Add(stats.balance, delta)
		if postAmt.Cmp(big.NewInt(0)) == 0 && amt.Cmp(big.NewInt(0)) > 0 {
```

Reply:
good point :+1: in that case, you could do

```suggestion
		delta := new(big.Int).Sub(postAmt, amt)
		stats.balance.Add(stats.balance, delta)
		if postAmt.Cmp(big.NewInt(0)) == 0 && amt.Cmp(big.NewInt(0)) > 0 {
```

which should be clearer?